src/platform/stm32f1/rcc.c

```diff
@@ -80,5 +83,136 @@ void rcc_init(enum stm32f1_external_clock_value clock)
+			break;
+		case RCC_CFGR_HPRE_DIV16:
+			rcc_clock_tree.ahb_clock_freq = rcc_clock_tree.sys_clock_freq >> 4;
+			break;
+		case RCC_CFGR_HPRE_DIV64:
+			rcc_clock_tree.ahb_clock_freq = rcc_clock_tree.sys_clock_freq >> 6;
+			break;
+		case RCC_CFGR_HPRE_DIV128:
+			rcc_clock_tree.ahb_clock_freq = rcc_clock_tree.sys_clock_freq >> 7;
+			break;
+		case RCC_CFGR_HPRE_DIV256:
+			rcc_clock_tree.ahb_clock_freq = rcc_clock_tree.sys_clock_freq >> 8;
+			break;
+		case RCC_CFGR_HPRE_DIV512:
+			rcc_clock_tree.ahb_clock_freq = rcc_clock_tree.sys_clock_freq >> 9;
+			break;
+	}
+
+	/* Determine apb bus 1 clock */
+	switch (RCC->CFGR & RCC_CFGR_PPRE1) {
+		case RCC_CFGR_PPRE1_DIV1:
+			rcc_clock_tree.apb1_clock_freq = rcc_clock_tree.ahb_clock_freq;
+			break;
+		case RCC_CFGR_PPRE1_DIV2:
+			rcc_clock_tree.apb1_clock_freq = rcc_clock_tree.ahb_clock_freq >> 1;
+			break;
+		case RCC_CFGR_PPRE1_DIV4:
+			rcc_clock_tree.apb1_clock_freq = rcc_clock_tree.ahb_clock_freq >> 2;
+			break;
+		case RCC_CFGR_PPRE1_DIV8:
+			rcc_clock_tree.apb1_clock_freq = rcc_clock_tree.ahb_clock_freq >> 3;
+			break;
+		case RCC_CFGR_PPRE1_DIV16:
+			rcc_clock_tree.apb1_clock_freq = rcc_clock_tree.ahb_clock_freq >> 4;
+			break;
+	}
+
+	/* Determine apb bus 2 clock */
+	switch (RCC->CFGR & RCC_CFGR_PPRE2) {
+		case RCC_CFGR_PPRE2_DIV1:
+			rcc_clock_tree.apb2_clock_freq = rcc_clock_tree.ahb_clock_freq;
+			break;
+		case RCC_CFGR_PPRE2_DIV2:
+			rcc_clock_tree.apb2_clock_freq = rcc_clock_tree.ahb_clock_freq >> 1;
+			break;
+		case RCC_CFGR_PPRE2_DIV4:
+			rcc_clock_tree.apb2_clock_freq = rcc_clock_tree.ahb_clock_freq >> 2;
+			break;
+		case RCC_CFGR_PPRE2_DIV8:
+			rcc_clock_tree.apb2_clock_freq = rcc_clock_tree.ahb_clock_freq >> 3;
+			break;
+		case RCC_CFGR_PPRE2_DIV16:
+			rcc_clock_tree.apb2_clock_freq = rcc_clock_tree.ahb_clock_freq >> 4;
+			break;
+	}
+
+	/* Determine apb bus 1 tim clock */
+	if ((RCC->CFGR & RCC_CFGR_PPRE1) == RCC_CFGR_PPRE1_DIV1)
+		rcc_clock_tree.apb1_tim_clock_freq = rcc_clock_tree.apb1_clock_freq;
+	else
+		rcc_clock_tree.apb1_tim_clock_freq = rcc_clock_tree.apb1_clock_freq >> 1;
+
+	/* Determine apb bus 2 tim clock */
+	if ((RCC->CFGR & RCC_CFGR_PPRE2) == RCC_CFGR_PPRE2_DIV1)
+		rcc_clock_tree.apb2_tim_clock_freq = rcc_clock_tree.apb2_clock_freq;
+	else
+		rcc_clock_tree.apb2_tim_clock_freq = rcc_clock_tree.apb2_clock_freq >> 1;
+
+	/* Determine adc clock */
+	switch (RCC->CFGR & RCC_CFGR_ADCPRE) {
+		case RCC_CFGR_ADCPRE_DIV2:
+			rcc_clock_tree.adc_clock_freq = rcc_clock_tree.apb2_clock_freq >> 1;
+			break;
+		case RCC_CFGR_ADCPRE_DIV4:
+			rcc_clock_tree.adc_clock_freq = rcc_clock_tree.apb2_clock_freq >> 2;
+			break;
+		case RCC_CFGR_ADCPRE_DIV6:
+			rcc_clock_tree.adc_clock_freq = rcc_clock_tree.apb2_clock_freq / 6;
+			break;
+		case RCC_CFGR_ADCPRE_DIV8:
+			rcc_clock_tree.adc_clock_freq = rcc_clock_tree.apb2_clock_freq >> 3;
+			break;
+	}
+}
+
+struct rcc_clock_tree_t rcc_clock_tree_get()
```

Comment:
`rcc_get_clock_tree`

But is this actually needed? Especially since we return a copy of the struct, not a pointer, so it can't be modified?

Reply:
And if we do actually want to make this accessible why not just making it visible to the other modules. In which case, don't make it static as I suggested below.

Reply:
not following? what is not need the function to return the struct?

Reply:
I was asking if we actually need to access the clock tree outside `rcc.c`. And if so, do we need to modify it?

Reply:
what are you suggesting? these values will be used in other functions like spi init to calculate prescalers, timers, etc

Reply:
they don't to modify them

Reply:
That is what I was asking.

I agree with the current approach then :blush:

Can you just rename the function to `rcc_get_clock_tree`? Seems more natural and easier to read.

Reply:
sure